---
title: "data_preprocessing"
author: "ddedesener"
date: "03/05/22"
output:
 md_document:
    variant: markdown_github
always_allow_html: true
---
## Introduction
In this script, visualization of enriched pathway which include both altered genes and metabolites is performed

## Setup
```{r setup, warning=FALSE, message=FALSE}
# check if libraries are already installed > otherwise install it
if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
if(!"rstudioapi" %in% installed.packages()) BiocManager::install("rstudioapi")
if(!"RCy3" %in% installed.packages()) BiocManager::install("RCy3")
if(!"rWikiPathways" %in% installed.packages()) BiocManager::install("rWikiPathways")
if(!"RColorBrewer" %in% installed.packages()) BiocManager::install("RColorBrewer")

#load libraries
library(rstudioapi)
library(RCy3)#connect cytoscape via R
library(rWikiPathways)#to get pathways from WikiPathways
library(RColorBrewer)#to manage colors with R

# set your working environment to the location where your current source file is saved into.
setwd(dirname(rstudioapi::getSourceEditorContext()$path))
```


## Import pathway
```{r}
#make sure to launch cytoscape, if you get CyREST error you need to relaunch cytoscape
cytoscapePing()

#close all opened session before starting
closeSession(FALSE)# clears all session data without saving

#pathway IDs
pathway.id <- "WP4723"# omega-3/omega-6 fatty acid synthesis pathway is one of the enriched pathways in our study
#so we will visualize the pathway in cytoscape

##### STEP-1 Open three enriched pathways #########
#import pathways as network in cytoscape
RCy3::commandsRun(paste0('wikipathways import-as-pathway id=',pathway.id )) 

#read gene data table from the file
geneData <- read.delim("data/geneData")
#load data to the pathway in cytoscape
loadTableData(table = "node", data = dataset, data.key.column = "ENSEMBL.ID", table.key.column = "Ensembl")
#read mbx data
mbxData <- read.delim("data/mbxData")
#load mbx data to the pathway in cytoscape
loadTableData(table = "node", data = mbxData, data.key.column = "CHEBI", table.key.column = "ChEBI")

##visual style##
RCy3::copyVisualStyle("default","pathwayStyle")
RCy3::setNodeLabelMapping("name", style.name="pathwayStyle")
RCy3::lockNodeDimensions(TRUE, style.name="pathwayStyle")
#assign different shapes based on data type
RCy3::setNodeShapeMapping('data.type', c('transcriptomics','metabolomics'), c("ellipse","rectangle"), style.name="pathwayStyle")
#chnage node sizes
RCy3::setNodeSizeMapping('data.type', c('transcriptomics','metabolomics'), c(40,25), mapping.type = "d", style.name = "pathwayStyle")

#color nodes like transcriptomics as pink, metabolomics yellow
setNodeColorMapping("data.type", c('transcriptomics','metabolomics'), c("#f5cbec","#f5ee31"), mapping.type = 'd', default.color = "#c6c6cf", style.name = "pathwayStyle")

#set node border color for showing data type
setNodeBorderColorMapping("data.type",c('transcriptomics','metabolomics'),c("#f5cbec","#f5ee31"),style.name="pathwayStyle")

#following codes are running but it applies the rule to all nodes
log2FCgenes = getTableColumns('node','log2FoldChange')  
min.logFC = min(log2FCgenes[,1],na.rm=TRUE)
max.logFC = max(log2FCgenes[,1],na.rm=TRUE)
data.values = c(min.logFC,0,max.logFC)

display.brewer.all(length(data.values), colorblindFriendly=TRUE, type="div") # div,qual,seq,all
node.colors <- c(rev(brewer.pal(length(data.values), "RdBu")))

setNodeColorMapping('log2FoldChange', data.values, node.colors, style.name="pathwayStyle")


RCy3::setVisualStyle("pathwayStyle")
RCy3::toggleGraphicsDetails()


```

