{
    "cells": [
        {
            "cell_type": "markdown",
            "metadata": {},
            "source": [
                "## Introduction \n",
                "In this workflow, differential gene expression analysis will be performed on host-transcriptomics data.\n",
                "\n",
                "## R environment setup\n",
                "First, install all required packages.\n"
            ]
        },
        {
            "cell_type": "code",
            "execution_count": null,
            "metadata": {},
            "outputs": [],
            "source": [
                "# check if BioCmanager libraries are already installed > otherwise install it\n",
                "if(!requireNamespace(\"BiocManager\", quietly = TRUE)) install.packages(\"BiocManager\",repos = \"http://cran.us.r-project.org\")\n",
                "if(!\"baySeq\" %in% installed.packages()) BiocManager::install(\"baySeq\")\n",
                "if(!\"DESeq2\" %in% installed.packages()) BiocManager::install(\"DESeq2\")\n",
                "if(!\"edgeR\" %in% installed.packages()) BiocManager::install(\"edgeR\")\n",
                "if(!\"bioDist\" %in% installed.packages()) BiocManager::install(\"bioDist\")\n",
                "if(!\"biomaRt\" %in% installed.packages()) BiocManager::install(\"biomaRt\")\n",
                "if(!\"dplyr\" %in% installed.packages()) BiocManager::install(\"dplyr\")\n",
                "if(!\"magrittr\" %in% installed.packages()) BiocManager::install(\"magrittr\")\n",
                "if(!\"EnhancedVolcano\" %in% installed.packages()) BiocManager::install(\"EnhancedVolcano\")\n",
                "\n",
                "#Regular R packages:\n",
                "if(!\"ggplot2\" %in% installed.packages()){install.packages(\"ggplot2\")}\n",
                "if(!\"limma\" %in% installed.packages()){install.packages(\"limma\")}\n",
                "#if(!\"R2HTML\" %in% installed.packages()){install.packages(\"R2HTML\")}\n",
                "\n",
                "#load packages\n",
                "library(baySeq)\n",
                "library(DESeq2)\n",
                "library(edgeR)\n",
                "library(bioDist)\n",
                "library(biomaRt)\n",
                "library(dplyr)\n",
                "library(magrittr)\n",
                "library(EnhancedVolcano)\n",
                "library(ggplot2)\n",
                "library(limma)\n",
                "#library(R2HTML)\n",
                "\n",
                "# set working environment to the location where current source file is saved into.\n",
                "#setwd(dirname(rstudioapi::getSourceEditorContext()$path))\n",
                "#include some functions adapted from ArrayAnalysis.org scripts\n",
                "source(\"functions_ArrayAnalysis_v2.R\")\n"
            ]
        },
        {
            "cell_type": "markdown",
            "metadata": {},
            "source": [
                "## Data Preparations\n",
                "The following section will prepare the input data to be used in the analysis\n"
            ]
        },
        {
            "cell_type": "code",
            "execution_count": null,
            "metadata": {},
            "outputs": [],
            "source": [
                "#set wd one directory below to reach the input data\n",
                "setwd('..')\n",
                "WORK.DIR <- getwd()\n",
                "\n",
                "#Obtain data from step 1\n",
                "htxCount <- read.csv(\"1-data_preprocessing/output/htxCount.csv\")\n",
                "sampleLabels <- read.csv(\"1-data_preprocessing/output/sampleLabels.csv\", header=FALSE)\n",
                "\n",
                "# Set Working Directory back to current folder\n",
                "#setwd(dirname(rstudioapi::getSourceEditorContext()$path))\n",
                "setwd(\"2-differential_gene_expression_analysis\")\n",
                "WORK.DIR<- getwd()\n",
                "\n",
                "#checking which samples have all zero values across all genes\n",
                "#these sample should be removed otherwise there will be a problem when calculating estimate size factors\n",
                "idx <- which(colSums(htxCount) == 0)\n",
                "#CSMDRVXI MSM719ME  are samples which has all zero values for all genes, so we remove them\n",
                "htxCount <- htxCount[ , -idx]\n",
                "\n",
                "#removing same samples from sample labels metadata\n",
                "sampleLabels <- sampleLabels[-idx , ]\n",
                "#Set column one as rownames\n",
                "rownames(sampleLabels) <- sampleLabels[,1]\n",
                "sampleLabels <- sampleLabels[,-1]\n",
                "#add column names\n",
                "colnames(sampleLabels) <- c( \"sampleID\", \"biopsy_location\",\"disease\")\n",
                "#check whether sample names are in same order\n",
                "#all(colnames(htxCount) == rownames(sampleLabels2))\n",
                "\n",
                "#select only biopsy_location and disease columns\n",
                "sampleLabels<-sampleLabels[, c(2,3)]\n",
                "sampleLabels$disease <- relevel(factor(sampleLabels$disease),ref=\"nonIBD\")\n",
                "#add an experimental group variable to sampleLabels\n",
                "sampleLabels$group <- as.factor(paste(sampleLabels$disease,sampleLabels$biopsy_location,sep=\"_\"))\n"
            ]
        },
        {
            "cell_type": "markdown",
            "metadata": {},
            "source": [
                "## Filtering Steps\n",
                "We will apply some filtering process to filter out genes in the input data\n"
            ]
        },
        {
            "cell_type": "code",
            "execution_count": null,
            "metadata": {},
            "outputs": [],
            "source": [
                "#remove genes which have all zero values across all samples then start DE analysis\n",
                "nonzero <- rowSums(htxCount) > 0\n",
                "htxCount %<>% .[nonzero,]\n",
                "\n",
                "#############################CPM FILTERING#############################################\n",
                "#aveLogCPM function computes average log2 counts-per-million for each row of counts.\n",
                "#the below function is similar to log2(rowMeans(cpm(y, ...)))\n",
                "mean_log_cpm = aveLogCPM(htxCount)\n",
                "\n",
                "# We plot the distribution of average log2 CPM values to verify that our chosen presence threshold is appropriate. The distribution is expected to be bi modal, with a low-abundance peak representing non-expressed genes and a high-abundance peak representing expressed genes. The chosen threshold should separate the two peaks of the bi modal distribution. \n",
                "filter_threshold <- -1# we can try different threshold values\n",
                "#jpeg(file=\"avgLogCpmDist.jpeg\")#if you want to save the histogram uncomment the following command  \n",
                "ggplot() + aes(x=mean_log_cpm) +\n",
                "    geom_histogram(binwidth=0.2) +\n",
                "    geom_vline(xintercept=filter_threshold) +\n",
                "    ggtitle(\"Histogram of mean expression values\")\n",
                "#dev.off()#to save the plot to the file\n",
                "#Having chosen our threshold, lets pick the subset of genes whose average expression passes that threshold.\n",
                "keep_genes <- mean_log_cpm >= filter_threshold \n",
                "htxCount <- htxCount[keep_genes,]\n",
                "#dim(htxCount)#to check dimension of the data\n",
                "###############################################################################\n"
            ]
        },
        {
            "cell_type": "markdown",
            "metadata": {},
            "source": [
                "## Differential Gene Expression Analysis\n",
                "In the following section differential gene expression analysis will be performed using DESeq2 package\n"
            ]
        },
        {
            "cell_type": "code",
            "execution_count": null,
            "metadata": {},
            "outputs": [],
            "source": [
                "# First create a DESeqDataSet object\n",
                "#(non-intercept) statistical model based on the disease and biopsy_location, group column represent both of them \n",
                "dds <- DESeqDataSetFromMatrix(countData = htxCount, colData=sampleLabels, design= ~0 + group)\n",
                "\n",
                "#estimate the size factors\n",
                "#To perform the median of ratios method of normalization, DESeq2 has a single estimateSizeFactors() function that will generate size factors for us.\n",
                "dds <- estimateSizeFactors(dds)\n",
                "#normalize the data (here for Quality Control(QC) plotting)\n",
                "#QC plotting is optional\n",
                "norm <- counts(dds,normalize=TRUE)\n",
                "#create a 2logged data for original object (here for QC plotting)\n",
                "datlog <- log(htxCount+1,2)\n",
                "#create a 2logged norm object (here for QC plotting)\n",
                "normlog <- log(norm+1,2)\n",
                "#for QC remove genes that have not been measured in any sample in the experiment\n",
                "datlogQC <- datlog[rowSums(datlog)!=0,]\n",
                "normlogQC <- normlog[rowSums(normlog)!=0,]\n",
                "\n",
                "#create QC plots for raw data, colored by different variables\n",
                "factors <- c(\"disease\",\"biopsy_location\",\"group\")\n",
                "if(!dir.exists(\"QCraw\")) dir.create(\"QCraw\")\n",
                "setwd(paste(WORK.DIR,\"QCraw\",sep=\"/\"))\n",
                "png(\"sizefactors.png\")\n",
                "plot(sizeFactors(dds),type='h',lwd=5,ylim=c(0,max(sizeFactors(dds))),col=\"darkblue\")\n",
                "dev.off()\n",
                "createQCPlots(datlogQC, factors, Table=sampleLabels, normMeth=\"\", postfix=\"\")\n",
                "setwd(\"..\")\n",
                "\n",
                "#create QC plots for normalized data colored by different variables\n",
                "if(!dir.exists(\"QCnorm\")) dir.create(\"QCnorm\")\n",
                "setwd(paste(WORK.DIR,\"QCnorm\",sep=\"/\"))\n",
                "createQCPlots(normlogQC, factors, Table=sampleLabels, normMeth=\"DESeq\", postfix=\"\")\n",
                "setwd(\"..\")\n",
                "\n",
                "#sample MSM719M9 is an outlier remove it from dataset\n",
                "#sample HSM5FZAZ is an outlier remove it from dataset\n",
                "htxCount <- htxCount[,-match(c(\"MSM719M9\",\"HSM5FZAZ\"),colnames(htxCount))]\n",
                "sampleLabels <- sampleLabels[-match(c(\"MSM719M9\",\"HSM5FZAZ\"),rownames(sampleLabels)),]\n",
                "#doublecheck whether the order of the samples in sampleLabels and htxCount data still match\n",
                "#sum(rownames(sampleLabels) == colnames(htxCount))==dim(sampleLabels)[1]\n",
                "\n",
                "###REDO ALL STEPS AS ABOVE FOR FILTERED OUT DATASET\n",
                "dds <- DESeqDataSetFromMatrix(countData = htxCount, colData=DataFrame(sampleLabels), design= ~0 + group)\n",
                "dds <- estimateSizeFactors(dds)\n",
                "norm <- counts(dds,normalize=TRUE)\n",
                "datlog <- log(htxCount+1,2)\n",
                "normlog <- log(norm+1,2)\n",
                "datlogQC <- datlog[rowSums(datlog)!=0,]\n",
                "normlogQC <- normlog[rowSums(normlog)!=0,]\n",
                "#create QC plots for raw data, coloured by different variables\n",
                "if(!dir.exists(\"QCraw2\")) dir.create(\"QCraw2\")\n",
                "setwd(paste(WORK.DIR,\"QCraw2\",sep=\"/\"))\n",
                "png(\"sizefactors2.png\")\n",
                "plot(sizeFactors(dds),type='h',lwd=5,ylim=c(0,max(sizeFactors(dds))),col=\"darkblue\")\n",
                "dev.off()\n",
                "createQCPlots(datlogQC, factors, Table=sampleLabels, normMeth=\"\", postfix=\"\")\n",
                "setwd(\"..\")\n",
                "\n",
                "#create QC plots for normalized data coloured by different variables\n",
                "if(!dir.exists(\"QCnorm2\")) dir.create(\"QCnorm2\")\n",
                "setwd(paste(WORK.DIR,\"QCnorm2\",sep=\"/\"))\n",
                "createQCPlots(normlogQC, factors, Table=sampleLabels, normMeth=\"DESeq\", postfix=\"\")\n",
                "setwd(\"..\")\n",
                "\n",
                "################################################################################\n",
                "#######################statistical modelling####################################\n",
                "################################################################################\n",
                "#set directory for stat output\n",
                "if(!dir.exists(\"statsmodel\")) dir.create(\"statsmodel\")\n",
                "setwd(paste(WORK.DIR,\"statsmodel\",sep=\"/\"))\n",
                "#run differential analysis\n",
                "dds <- DESeq(dds)\n",
                "cont.matrix <- makeContrasts(\n",
                "  #CD disease on ileum and rectum \n",
                "  CD_Ileum_vs_nonIBD_Ileum   = groupCD_Ileum - groupnonIBD_Ileum,\n",
                "  CD_Rectum_vs_nonIBD_Rectum = groupCD_Rectum - groupnonIBD_Rectum,\n",
                "  #UC disease on ileum and rectum\n",
                "  UC_Ileum_vs_nonIBD_Ileum   = groupUC_Ileum - groupnonIBD_Ileum,\n",
                "  UC_Rectum_vs_nonIBD_Rectum = groupUC_Rectum - groupnonIBD_Rectum,\n",
                "  #UC and CD disease comparison\n",
                "  UC_Ileum_vs_CD_Illeum     = groupUC_Ileum - groupCD_Ileum,\n",
                "  UC_Rectum_vs_CD_Rectum    = groupUC_Rectum - groupCD_Rectum,\n",
                "  #biopsy location comparisons\n",
                "  CD_Rectum_vs_CD_Ileum     = (groupCD_Rectum - groupnonIBD_Rectum) - (groupCD_Ileum - groupnonIBD_Ileum),\n",
                "  UC_Rectum_vs_UC_Ileum     = (groupUC_Rectum - groupnonIBD_Rectum) - (groupUC_Ileum - groupnonIBD_Ileum),\n",
                "  levels = resultsNames(dds)\n",
                ")\n",
                "#extract resulting contrasts based on the model, and save those in a table; also save some graphical representations\n",
                "#the function results() is called from within the saveStatOutputDESeq2 function to compute the contrasts\n",
                "files <- saveStatOutputDESeq2(cont.matrix,dds,postfix=\"\",annotation=NULL)\n",
                "#create summary table of the contrast results\n",
                "createPvalTab(files,postfix=\"\",namePVal=\"pvalue\",nameAdjPVal=\"padj\",nameFC=\"FoldChange\",nameLogFC=\"log2FoldChange\",html=TRUE)\n",
                "setwd(\"..\")\n"
            ]
        },
        {
            "cell_type": "markdown",
            "metadata": {},
            "source": [
                "## Enhanced Volcano Plots Visualization\n",
                "Differential gene expression analysis results is visualized by volcano plots\n"
            ]
        },
        {
            "cell_type": "code",
            "execution_count": null,
            "metadata": {},
            "outputs": [],
            "source": [
                "#set the wd to the result folder statsmodel\n",
                "setwd(paste(WORK.DIR,\"statsmodel\",sep=\"/\"))\n",
                "readFilePath <- paste(WORK.DIR,\"statsmodel\",sep=\"/\")\n",
                "# create an empty list\n",
                "plot_list = list()\n",
                "for(i in 1:length(files))\n",
                "{\n",
                "  #read file\n",
                "  splitted <- strsplit(files[i], split = \"_\")[[1]]\n",
                "  title <- splitted [2:6]\n",
                "  title <- paste (splitted[2],splitted[3],splitted[4],splitted[5],splitted[6],sep=\" \") \n",
                "  title <- (strsplit(title, split = \"\\\\.\")[[1]])[1]\n",
                "  tab <- read.delim(paste(readFilePath, files[i],sep=\"/\"),header=TRUE,as.is=TRUE)  \n",
                "  p<-EnhancedVolcano(tab , lab = tab$X, labSize = 3, title = title, \n",
                "                     x = 'log2FoldChange', y = 'pvalue', pCutoff = 0.05, FCcutoff = 0.58)\n",
                "  plot_list[[i]] = p\n",
                "}\n",
                "#path of the output folder\n",
                "outFolder <- paste(WORK.DIR,\"volcano_plots\",sep=\"/\")\n",
                "#create folder if doesnt exist\n",
                "if(!dir.exists(outFolder)) dir.create(outFolder)\n",
                "\n",
                "for(i in 1:length(files)) {\n",
                "    file_name = paste(outFolder,\"/\",files[i],\".png\",sep=\"\")\n",
                "    png(file_name)\n",
                "    print(plot_list[[i]])\n",
                "    dev.off()\n",
                "}  \n"
            ]
        },
        {
            "cell_type": "markdown",
            "metadata": {},
            "source": "##Print session info and remove large datasets:\n"
        },
        {
            "cell_type": "code",
            "execution_count": null,
            "metadata": {},
            "outputs": [],
            "source": [
                "##Print session info:\n",
                "sessionInfo()\n",
                "\n",
                "##Remove data objects which are not needed for further processing:\n",
                "rm(list=setdiff(ls(), \"tab\"))\n"
            ]
        },
        {
            "cell_type": "markdown",
            "metadata": {},
            "source": "### Last, we create a Jupyter notebook from this script\n"
        },
        {
            "cell_type": "code",
            "execution_count": null,
            "metadata": {},
            "outputs": [],
            "source": [
                "#Jupyter Notebook file\n",
                "if(!\"devtools\" %in% installed.packages()) BiocManager::install(\"devtools\")\n",
                "devtools::install_github(\"mkearney/rmd2jupyter\", force=TRUE)\n",
                "library(devtools)\n",
                "library(rmd2jupyter)\n",
                "#setwd(dirname(rstudioapi::getSourceEditorContext()$path))\n",
                "rmd2jupyter(\"DEanalysis.Rmd\")\n"
            ]
        }
    ],
    "metadata": {
        "anaconda-cloud": "",
        "kernelspec": {
            "display_name": "R",
            "langauge": "R",
            "name": "ir"
        },
        "language_info": {
            "codemirror_mode": "r",
            "file_extension": ".r",
            "mimetype": "text/x-r-source",
            "name": "R",
            "pygments_lexer": "r",
            "version": "3.4.1"
        }
    },
    "nbformat": 4,
    "nbformat_minor": 1
}
